Core/clim-basic/geometry/region-predicates.lisp

```diff
@@ -83,21 +134,80 @@
                (wn 0))
            (map-over-polygon-segments
             (lambda (x1 y1 x2 y2)
+              ;; Algorithm is not predictible for polygon edges - we
+              ;; need to test for them explicitly. -- jd 2019-09-27
+              (when (segment-contains-point-p x1 y1 x2 y2 x y)
+                (return-from region-contains-position-p t))
               (if (<= y1 y)
                   (when (and (> y2 y)
                              (> (is-left x1 y1 x2 y2 x y) 0))
                     (incf wn))
                   (when (and (<= y2 y)
                              (< (is-left x1 y1 x2 y2 x y) 0))
                     (decf wn))))
-            self)
+            region)
            (not (zerop wn))))))
 
+(defmethod region-contains-position-p ((self rectangle) x y)
+  (multiple-value-bind (x1 y1 x2 y2)
+      (rectangle-edges* self)
+    (and (coordinate-between* x1 x x2)
+         (coordinate-between* y1 y y2))))
+
+(defmethod region-contains-position-p ((self ellipse) x y)
+  (nest
+   (multiple-value-bind (cx cy) (ellipse-center-point* self))
+   (let ((alpha (or (ellipse-start-angle self) 0))
+         (omega (or (ellipse-end-angle self) (* 2 pi)))
+         (polar->screen (polar->screen self)))
+     (and (angle-contains-point-p alpha omega (- x cx) (- y cy))
```

Comment:
This does lots of unnecessary work if start and end angle are null. Conversely, if the angle test fails, the `polar->screen` call is wasted. I suggest making local functions `angle-contains-p` for the angle test and `position-contains-p` for the position test and organizing the body like
```lisp
(and (or (not start-angle)
         (angle-contains-p start-angle (or end-angle 2pi)))
     (position-contains-p))
```

Reply:
fixed (I've used slightly different code). if start-angle is not null then end-angle is not null neither (specified behavior) so I've skipped (or end-angle 2pi).